pkg/collector/corechecks/cluster/kubernetes_state_label_joins.go

```diff
@@ -0,0 +1,208 @@
+
+	for key, joinsConfig := range config {
+		if len(joinsConfig.LabelsToMatch) > 0 {
+			metricsToJoin[key] = metricToJoin{
+				config: joinsConfig,
+				tree:   newInnerNode(),
+			}
+		} else {
+			metricsToJoin[key] = metricToJoin{
+				config: joinsConfig,
+				tree:   newLeafNode(),
+			}
+		}
+	}
+
+	return &labelJoiner{
+		metricsToJoin: metricsToJoin,
+	}
+}
+
+func newInnerNode() *node {
+	return &node{
+		labelValues: make(map[string]*node),
+		labelsToAdd: nil,
+	}
+}
+
+func newLeafNode() *node {
+	return &node{
+		labelValues: nil,
+		labelsToAdd: nil,
+	}
+}
+
+func (lj *labelJoiner) insertMetric(metric ksmstore.DDMetric, config *JoinsConfig, tree *node) {
+	current := tree
+	nbLabelsToMatch := len(config.LabelsToMatch)
+	for i, labelToMatch := range config.LabelsToMatch {
+		labelValue, found := metric.Labels[labelToMatch]
+		if !found {
+			return
+		}
+
+		child, found := current.labelValues[labelValue]
+		if !found {
+			if i < nbLabelsToMatch-1 {
+				child = newInnerNode()
+			} else {
+				child = newLeafNode()
+			}
+			current.labelValues[labelValue] = child
+		}
+
+		current = child
+	}
+
+	if config.GetAllLabels {
+		if current.labelsToAdd == nil {
+			current.labelsToAdd = make([]label, 0, len(metric.Labels)-len(config.LabelsToMatch))
+		}
+
+		for labelName, labelValue := range metric.Labels {
+			isALabelToMatch := false
+			for _, labelToMatch := range config.LabelsToMatch {
+				if labelName == labelToMatch {
+					isALabelToMatch = true
+					break
+				}
+			}
+			if !isALabelToMatch {
+				current.labelsToAdd = append(current.labelsToAdd, label{labelName, labelValue})
+			}
+		}
+	} else {
+		if current.labelsToAdd == nil {
+			current.labelsToAdd = make([]label, 0, len(config.LabelsToGet))
+		}
+
+		for _, labelToGet := range config.LabelsToGet {
+			labelValue, found := metric.Labels[labelToGet]
+			if found {
+				current.labelsToAdd = append(current.labelsToAdd, label{labelToGet, labelValue})
+			}
+		}
+	}
+}
+
+func (lj *labelJoiner) insertFamily(metricFamily ksmstore.DDMetricsFam) {
+	metricToJoin, found := lj.metricsToJoin[metricFamily.Name]
+	if !found {
+		log.Error("BUG in label joins")
+		return
+	}
+
+	for _, metric := range metricFamily.ListMetrics {
+		lj.insertMetric(metric, metricToJoin.config, metricToJoin.tree)
+	}
+}
+
+func (lj *labelJoiner) insertFamilies(metrics map[string][]ksmstore.DDMetricsFam) {
+	for _, metricsList := range metrics {
+		for _, metricFamily := range metricsList {
+			lj.insertFamily(metricFamily)
+		}
+	}
+}
+
+func (lj *labelJoiner) getLabelsToAddOne(inputLabels map[string]string, config *JoinsConfig, tree *node, labelsToAdd *[]label) {
+	node := tree
+	for _, labelToMatch := range config.LabelsToMatch {
+		labelValue, found := inputLabels[labelToMatch]
+		if !found {
+			return
+		}
+
+		node, found = node.labelValues[labelValue]
+		if !found {
+			return
+		}
+	}
+
+	*labelsToAdd = append(*labelsToAdd, node.labelsToAdd...)
```

Comment:
I now see where the discussion about slices by reference came about 😄 Another alternative could be to invent a struct that holds the state and implements the label append on that state and then return the result of that. Not sure if that would yield better performance though.

Reply:
I made the same proposition about a extra struct to own the slice :)